llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -3442,6 +3442,9 @@ static SDValue combineCarryDiamond(SelectionDAG &DAG, const TargetLowering &TLI,
     return SDValue();
   if (Opcode != ISD::UADDO && Opcode != ISD::USUBO)
     return SDValue();
+  // Guarantee identical type of CarryOut
+  if (N->getValueType(0) != Carry0.getValue(1).getValueType())
```

Comment:
Most of this code implicitly assumes we're returning == MVT::i1, so maybe:
```cpp
if (N->getValueType(0) != MVT::i1 || 
    Carry0.getValue(1).getValueType() != MVT::i1 ||
    Carry1.getValue(1).getValueType() != MVT::i1)
  return SDValue();
```

Reply:
I don't see how this is implicitly assuming i1? I can see implicitly assuming the uaddo output and input bools are the same